### ISSUE TITLE
Make reader and author registration optional / opt-in

### ISSUE THREAD

**Issue by jmacgreg** [pkp/pkp-lib#3600]
Right now it looks like, when a user registers for a journal (OJS 3) they are automatically registered as authors and readers, with an option to also register as a reviewer. If there is an option to register for multiple different roles, these should be explicitly listed and available as opt-in only.

**Comment by asmecher** [comment]
This was an intentional change made with the front-end refactor; see https://github.com/pkp/pkp-lib/issues/1443#issuecomment-219802606. Does GDPR prohibit us from pre-checking these checkboxes? (Tagging @NateWr)

**Comment by ajnyga** [comment]
I do not think it prohibits us from suggesting the selections as long as you can remove the selection easily.

**Comment by NateWr** [comment]
**AUDIBLE GROAN**

**Comment by jmacgreg** [comment]
Hey @NateWr here comes some more cause to groan :-) The multi-role nature of the role system in OJS makes it a little difficult to manage 0, 1 or 2+ roles in the registration list, but I think I've got a reasonable default. Thoughts on the following screenshots? 

No author roles: nothing appears. 
<img width="1440" alt="register_noauthorroles" src="https://user-images.githubusercontent.com/201446/39709223-b1b97a9e-51ef-11e8-8b76-28b97ef22435.png">

One author role: appears with a legend + confirmation checkbox. 
<img width="1440" alt="register_oneauthorrole" src="https://user-images.githubusercontent.com/201446/39709222-b1a4a128-51ef-11e8-89a6-04a2ca3e3a31.png">

Multiple author roles: each appears with a legend + confirmation checkbox. Ideally I'd like to group multiple role types under a single legend, but that will require a more involved overhaul. 
<img width="1440" alt="register_multipleauthorroles" src="https://user-images.githubusercontent.com/201446/39709221-b192c084-51ef-11e8-8db9-9c89c0a141a9.png">

**Comment by jmacgreg** [comment]
I'd eventually like to expand the roles to include a description for each role, which could then be displayed on registration and in the user's profile. But that would be another issue entirely.

**Comment by ajnyga** [comment]
I think both suggestions to revise the language around the roles is very good! Even editors seem to mix "Reader" with subscriptions from time to time. 

I also agree that #2297 is a fairly straight-forward solution for assigning the author role. Especially in a multijournal installations where there are users around with no roles in some context. 

(We could even have a block plugin for journals that would show a "notify me of new publications" button for users that are logged in. If the user is logged in and already has that role, the button shows an option to unsubscribe from those notifications, which basically drops the reader role. This would probably be a lot easier way of adding/removing the reader role within a context if you compare it to editing your profile settings.)

**Comment by ajnyga** [comment]
@NateWr @jmacgreg 
Realized just now (edited the above message) that you can not have a text like "Yes, I would like to be notified of new publications." and tie that to the Reader role, because all the roles receive the notifications. It would lead into situations where someone does not check that but still receives those notifications.

My suggestion is that we:
- Change the system so that "public" notifications (new issue, announcement) **only** go to users with the `Reader` role.
- We rename the whole `Reader` role to something that describes the role better. (have a lot of good suggestions in Finnish :P)
- In the upgrade script to 3.1.1.1 we add the `Reader` role to **all** users with some role within a context.
- Change the notification settings in the profile settings so that disabling/enabling the public notifications (new issue, announcement) basically means adding/removing the Reader role.
- And we now even have a simple way of providing the notification emails with a link "I do not want to receive these kinds of notifications anymore form $context"

**Comment by jmacgreg** [comment]
Hi all, 

I strongly agree with @NateWr's suggestions. I can remove the behind-the-scenes author opt-in options from the registration page in deference to #2297. I can also make the suggested text changes. I don't think we need to explicitly state anywhere that users can opt out elsewhere. 

@ajnyga I really like your suggestions too, and I think it's imperative to limit "public" notifications from going out to everyone - they should only go out to readers as you suggest. I like your points here, but they are beyond my ability to code I think. Are you able to do these for 3.1.1-1? Should they be filed in a separate bug?

**Comment by ajnyga** [comment]
I think I could do a pr along the suggestion above, but maybe @asmecher or @NateWr could first comment on that. It is a fairly big change to the way the system works at the moment and I am not sure if you actually want to tie the notifications to a specific role. Do we actually need the Reader role or just control over the notifications?

**Comment by NateWr** [comment]
Ok, I will work on the review for #2297 so we can get that merged in for 3.1.1-1.

> I think it's imperative to limit "public" notifications from going out to everyone - they should only go out to readers as you suggest.

Who is getting harmed here, and would it be onerous for them to opt-out of these notifications? I can see how the Reader role offers a ready-made solution for this problem. But I'm not sure about pairing an entire role to a specific feature -- and one that they can opt out of. I think we'll be boxing ourselves in a bit. The only thing that jumps to mind right now is journals that require users to be logged in to view articles (is that a thing or am I making it up?). If someone opts out of notifications, and we then remove the role, they may not be able to access articles.

I think we probably need to tie this opt-in during the registration process to a setting in the Profile > Notifications tab. (I looked, but didn't see an option for opting out of publication notifications.) Anyone can opt in or out, regardless of what role they have in the system.

Even if they don't check the notifications box, they should get enrolled as a Reader in lieu of any other role selection (the only remaining one for self-registration being Reviewer). It's hard for me to see a circumstance where this shouldn't be permitted on privacy grounds. If someone registers a user account, and does not select a Reviewer role, they would still have a reasonable expectation of having _some_ role in the system, and Reader would be that role. 

> In the upgrade script to 3.1.1.1 we add the Reader role to all users with some role within a context.

I'm in favor of this. Any user with an account that should be attached to a context in any way should be treated as a Reader in that context. Otherwise, it's hard to see the role making much sense.

**Comment by NateWr** [comment]
(Agree that @asmecher or @bozana should look at this too, so we're not missing anything here.)

**Comment by ajnyga** [comment]
If we do not attach the notification functionality to the reader role, then where do we need that role? 

We could just have the option of enabling/disabling the public notifications?

**Comment by ajnyga** [comment]
I am actually not sure if we can assign a new role automatically to all users. I mean we can if that happens before GDPR but most journals will upgrade after that date and GDPR explicitly denies exactly this kind of automation. I originally suggested that as part of the solution where we would tie the notifications to the Reader role. But as @NateWr made clear above, this is probably not a good idea. So I would forget about the upgrade automation.

I am still having a hard time seeing where the Reader role is actually needed. Even with the situation where a journal is limiting the visibility of articles for registered users, does that user actually need a role within that context, or is it enough that she a registered user in that OJS installation?

**Comment by NateWr** [comment]
> GDPR explicitly denies exactly this kind of automation

IAONAL (I Am Obviously Not A Lawyer), but if assigning the Reader role is simply a refactor of the code base, implies no additional privileges in the platform, and collects no additional information, then in what way would it be blocked by GDPR? 

I am struggling to see how this this would be different from adding an entry to the settings table with a key "userCanAccessContextId", which would simply be permanently storing a piece of data which we normally calculate at run time. And if that's not allowed, then how could we take any normal action that comes with maintaining software, like changing "firstName" to "givenName", or migrating salutation and suffix into a `prefferedPublicName` field, or renaming a userGroup from `Reader` to `Subscriber`?

I would understand if assigning the Reader role were to move a user into a bucket of notifications/features that they did not opt into. But if a user has an existing role in a context, they have effectively already opted into the Reader role. The internal data is just not structured this way at the moment.

**Comment by jmacgreg** [comment]
I think we're all in agreement that using the Reader role to dictate whether someone should be getting notifications is a bad idea - all users should be able to opt in/out of those notifications, and this should be clearly part of registration and part of the user profile. This has already been filed at https://github.com/pkp/pkp-lib/issues/2805, but against OJS 3.2. I know it's a big ask, but is there any chance this can be considered for 3.1.1-1? And can this issue now be closed?

**Comment by NateWr** [comment]
> all users should be able to opt in/out of those notifications, and this should be clearly part of registration and part of the user profile. This has already been filed at #2805, but against OJS 3.2. I know it's a big ask, but is there any chance this can be considered for 3.1.1-1?

It seems to me like #2805 is a requirement for GDPR anyway, right?

> (I think users can access articles also without a role if a journal requires user to be logged in for reading)

If so, that should probably be changed. A journal has a reasonable expectation the this setting will require them to be part of their Reader statistics. There needs to be some way for a logged in user to easily opt into a Reader role in a context before accessing an article.

> all roles are paired with specific features/abilities. You need the Author role to submit, Reviewer role to review, why not need the Reader role to receive public notifications?

You're right, but for a basket of permissions, not just a feature. An Author role is needed to submit, to access a submissions list, to be assigned to a submission, and so-on. My objection is to tying a role to a single feature, rather than to permission to access something.

> Even if a user is registered without a role, it should not create a problem because of #2297

I think we'll run into other problems for users without a registered role. For example, a Journal Manager won't have access to a user without any role. A user without any role in a context (except admin) is effectively excluded from the entire system (or should be), and should probably be removed as an inactive user. A group assignment is how we attach permissions to a user in a context. I don't want to re-use a system that was designed to do one thing to try and make it do another.

I'll look further into #2805 as well as @ajnyga's work on #2297. My sense is that #2805 is probably necessary for GDPR anyway, and even if it's not it's long overdue.

**Comment by NateWr** [comment]
> I think the solution here has to be based on two levels of privacy policies

Sorry, just seeing your edit now. Yes, I suppose you're probably right. I've long advocated for _not having_ the site-wide user registration, but that was overruled last time, I think so that Assistants can more easily register across a bunch of journals... But in a very large installation, this really doesn't make a lot of sense....

**Comment by ajnyga** [comment]
> If so, that should probably be changed. A journal has a reasonable expectation the this setting will require them to be part of their Reader statistics. There needs to be some way for a logged in user to easily opt into a Reader role in a context before accessing an article.

I am fairly sure, see https://github.com/pkp/ojs/blob/master/pages/article/ArticleHandler.inc.php#L316

We could just have something similar like with Authors in #2297 when someone want to read an article in such journal. But does the role have to be specifically the Reader role? Meaning that for example Reviewers do not automatically have that ability?

However, I think that only a few journals limit the access to articles like that so in multijournal installations the majority of readers that are registered to other contexts will not have the Reader role. 

If you think that we should encourage people to enroll as Readers if the are readers of the journal, then maybe the block plugin I suggested above could work. It could still have a button saying "Notify me of new issues and announcements". For users without the Reader role in that context it would assign the Reader role (tell the user about it) and would activate the public notifications. For users with the Reader role it would just activate the notifications. And with users with both the Reader role and the activated notifications, it would show a button to disable those notifications. This would be an easy way to opt-in to notifications and to the Reader role but would be compatible with the model you suggested.

What I am also saying is that we maybe do not want public notification settings for users without any role in the context, right? Because the role is what ties the user to the context policies.

But even with this solution, in multijournal installations we will always have situations where we have registered users with no roles reading the content of a context. So we can not trust that the readers are all Readers and build functionalities on that basis.

> Yes, I suppose you're probably right. I've long advocated for not having the site-wide user registration, but that was overruled last time, I think so that Assistants can more easily register across a bunch of journals... But in a very large installation, this really doesn't make a lot of sense....

We have 60 journals and I can imagine that the whole form looks just confusing for new users. I think that in site level, instead of making users choose the roles (and read all the context specific privacy policies that come with those roles!) we should just leave those users without roles and find ways (like above) to suggest roles within a context in a later stage when we actually want/offer the user to do something that requires a role. Like in #2297.
